- Description: cio: unexpected event in state machine leads to kernel panic.
  Symptom:     After a logical vary off of a networking chpid on which a device
               is in use, "dev_jumptable[0][1]==0" is printed and a BUG()
               occurs (only observed with pre-emptible kernel).
  Problem:     There is a window where an unsolicited interrupt may come in for
               a subchannel which has no longer an associated ccw_device, but for
               which the subchannel's intparm is still set, leading to the
               attempted delivery of an interrupt for a non-existent device.
  Solution:    Correctly disable the subchannel in ccw_device_nopath_notify,
               because after a logical vary off the subchannel is still existing
               but must not produce (unsolicited) interrupts.
  Problem-ID:  -
  Archs:       s390-31, s390-64

--- linux-2.5/drivers/s390/cio/device_fsm.c	7 Apr 2004 14:56:48 -0000	1.85
+++ linux-2.5/drivers/s390/cio/device_fsm.c	20 Apr 2004 11:54:31 -0000	1.85.2.1
@@ -440,12 +440,14 @@
 	if (!ret) {
 		if (get_device(&sch->dev)) {
 			/* Driver doesn't want to keep device. */
+			cio_disable_subchannel(sch);
 			device_unregister(&sch->dev);
 			sch->schib.pmcw.intparm = 0;
 			cio_modify(sch);
 			put_device(&sch->dev);
 		}
 	} else {
+		cio_disable_subchannel(sch);
 		ccw_device_set_timeout(cdev, 0);
 		cdev->private->state = DEV_STATE_DISCONNECTED;
 		wake_up(&cdev->private->wait_q);

